### PR TITLE
[Consensus] add streaming support to tonic network

### PR DESCRIPTION
**Pull request by mwtian**
## Description 

This is the first PR of the streaming implementation. Tonic streaming, and streaming support in `AuthorityService` are implemented.

But block propagation with tonic is still using unary RPC. It will switch to use streaming after adding client side logic for block subscription in future.

## Test Plan 

CI

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: consensus/core/src/authority_service.rs
Comment (mwtian) [inline consensus/core/src/authority_service.rs:1]: This file is mostly taken from authority_node.rs, except for move the body of `handle_send_block()` into `impl AuthorityService`, and adding `handle_subscribe_blocks()`.

File: consensus/core/src/network/tonic_network.rs [hunk 1]
```diff
@@ -115,11 +163,44 @@ impl NetworkClient for TonicClient {
...
+        let mut blocks = vec![];
+        let mut total_fetched_bytes = 0;
+        loop {
+            match stream.message().await {
+                Ok(Some(response)) => {
+                    for b in &response.blocks {
+                        total_fetched_bytes += b.len();
+                    }
+                    blocks.extend(response.blocks);
+                    if total_fetched_bytes > MAX_TOTAL_FETCHED_BYTES {
```
Comment (arun-koshy) [inline consensus/core/src/network/tonic_network.rs:180]: Curious if you think this is something that would be useful to pass in as part of the request so the server knows how much to send back. Then if we do exceed that threshold we know it is indeed byzantine behavior?

File: consensus/core/src/network/mod.rs [hunk 2]
```diff
@@ -29,13 +30,19 @@ pub(crate) mod epoch_filter;
...
+
 /// Network client for communicating with peers.
 ///
 /// NOTE: the timeout parameters help saving resources at client and potentially server.
 /// But it is up to the server implementation if the timeout is honored.
 /// - To bound server resources, server should implement own timeout for incoming requests.
 #[async_trait]
 pub(crate) trait NetworkClient: Send + Sync + 'static {
+    // Whether the network client streams blocks to subscribed peers.
+    const SUPPORT_STREAMING: bool;
```
Comment (arun-koshy) [inline consensus/core/src/network/mod.rs:44]: I may just be missing it but how is this used to gate streaming?

**Review by arun-koshy — APPROVED** [review]

File: consensus/core/src/network/mod.rs (hunk 2 above)
Comment (mwtian) [reply consensus/core/src/network/mod.rs:59]: Yeah this is unused for this PR, but will be used in the follow up to decide if new blocks are propagated with Broadcaster or another mechanism.

File: consensus/core/src/network/tonic_network.rs (hunk 1 above)
Comment (mwtian) [reply consensus/core/src/network/tonic_network.rs:181]: Definitely, this does seem like a more flexible approach. Adding a TODO to `NetworkClient::fetch_blocks()`.

File: consensus/core/src/network/tonic_network.rs [hunk 3]
```diff
@@ -29,18 +32,24 @@ use super::{
...
     context::Context,
     error::{ConsensusError, ConsensusResult},
     network::tonic_gen::consensus_service_server::ConsensusServiceServer,
+    Round,
 };
 
 const AUTHORITY_INDEX_METADATA_KEY: &str = "authority-index";
 
+// Maximum bytes size in a single fetch_blocks()response.
+const MAX_FETCH_RESPONSE_BYTES: usize = 4 * 1024 * 1024;
```
[the remark this answers is not in the document]
Comment (mwtian) [reply consensus/core/src/network/tonic_network.rs:49]: Yes, but I think it is better to keep the per-message size limit consistent across the valdiators. Adding a TODO.

File: consensus/core/src/network/tonic_network.rs [hunk 4]
```diff
@@ -519,4 +737,60 @@ mod test {
...
+                Duration::from_secs(5),
+            )
+            .await
+            .unwrap();
+        receive_stream_0
+            .enumerate()
+            .for_each_concurrent(None, |(i, item)| async move {
+                assert_eq!(item, block_for_round(client_0_round + i as Round + 1));
+            })
+            .await;
```
Comment (akichidis) [inline consensus/core/src/network/tonic_network.rs:783]: Maybe we want to confirm that this stream has indeed received the expected number of blocks? This stream might not return anything and the test will still pass

**Review by akichidis — APPROVED** [review]

File: consensus/core/src/dag_state.rs [hunk 5]
```diff
@@ -326,6 +326,27 @@ impl DagState {
         genesis_block.clone()
     }
 
+    /// Returns cached recent blocks from the input authority, starting from round `start`.
+    /// Number of blocks returned is limited by the number of blocks cached.
+    pub(crate) fn get_cached_blocks(
```
[the remark this answers is not in the document]
Comment (akichidis) [reply consensus/core/src/dag_state.rs:331]: Same as we discussed in the other methods where only data from the last GC round are guaranteed to be without gaps.

[the remark this answers is not in the document]
Comment (mwtian) [reply consensus/core/src/dag_state.rs:331]: We can talk offline. Only getting cached blocks is mentioned here. Also this function retrieves blocks for a single authority, which is guaranteed to be "continuous" but I think this is less relevant to causal completion.

**Pull request merged**